### PR TITLE
Add project-wide Metadata Items

### LINKED ISSUE #17: GCP: Project Metadata
> You can store SSH keys that can be used to connect to the VM instances of a project. Project-level keys are propagated to all VM instances that do not have their own SSH keys.

I could not find an API for this console feature. Seems like one should exist, perhaps buried in another bigger API? If one doesn't exist this ticket can be closed. 

ssh-keys

### PR DESCRIPTION
**Pull request by alex-lambrides**
Fixes #17: This PR adds project wide metadata items, which can be used to set the `ssh-keys` metadata item.

### CODE REVIEW

File: src/main/java/gyro/google/compute/ProjectMetadataItemResource.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+@Type("project-metadata-item")
+public class ProjectMetadataItemResource extends ComputeResource implements Copyable<Metadata.Items> {
+    private String key;
+    private String value;
+
+    /**
+     * The key of the metadata item. (Required)
+     */
+    @Required
+    public String getKey() {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ProjectMetadataItemResource.java]: This field has restrictions on what can be set, so you should add validation to it.

You can follow this:
https://github.com/perfectsense/gyro-google-provider/blob/feature/bucket/src/main/java/gyro/google/storage/BucketResource.java#L220-L244

File: src/main/java/gyro/google/compute/ProjectMetadataItemResource.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        try {
+            Compute.Projects.SetCommonInstanceMetadata metadataRequest =
+                client.projects().setCommonInstanceMetadata(getProjectId(), metadata);
+            Operation operation = metadataRequest.execute();
+
+            Operation.Error error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+        } catch (GoogleJsonResponseException je) {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ProjectMetadataItemResource.java]: Currently for duplicate keys the error response looks like
```
Invalid value for field 'resource': '{  "fingerprint": "KjSjVaX4e8I=",  "item": [{    "key": "ssh-keys",    "value": "deepanjan:ssh-rsa A...'. Metadata has duplicate keys: ssh-keys
```

For this particular case we could know that its a duplicate key, and throw a proper formatter error
Something along the lines of `Duplicate key: <Key_Name>`

File: src/main/java/gyro/google/compute/ProjectMetadataItemFinder.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        } catch (IOException ex) {
+            throw new GyroException(ex);
+        }
+    }
+
+    @Override
+    protected List<Metadata.Items> findGoogle(Compute client, Map<String, String> filters) {
+        try {
+            Metadata.Items item = client.projects().get(getProjectId()).execute().getCommonInstanceMetadata().getItems().stream()
+                .filter(r -> filters.get("name").equals(r.getKey()))
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ProjectMetadataItemFinder.java]: `name` should be `key` as that is the field in the class.

running a gyro file containing --> `@print: $(external-query google::project-metadata-item {key: 'ssh-keys'})` results in :
```
Error: Can't process the @print directive!

In compute/project-metadata-item.gyro on line 1 from column 1 to 72:
1: @print: $(external-query google::project-metadata-item {key: 'ssh-keys'})

Caused by: Error: Can't resolve external-query reference!

In compute/project-metadata-item.gyro on line 1 from column 9 to 72:
1: @print: $(external-query google::project-metadata-item {key: 'ssh-keys'})

Caused by: Unexpected error: java.lang.NullPointerException
	at gyro.google.compute.ProjectMetadataItemFinder.lambda$findGoogle$0(ProjectMetadataItemFinder.java:71)
	at gyro.google.compute.ProjectMetadataItemFinder.lambda$findGoogle$0(ProjectMetadataItemFinder.java:71)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1631)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at gyro.google.compute.ProjectMetadataItemFinder.findGoogle(ProjectMetadataItemFinder.java:72)
	at gyro.google.compute.ProjectMetadataItemFinder.findGoogle(ProjectMetadataItemFinder.java:41)
	at gyro.google.GoogleFinder.find(GoogleFinder.java:41)
	at gyro.core.reference.FinderReferenceResolver.resolve(FinderReferenceResolver.java:57)
	at gyro.core.scope.NodeEvaluator.visitReference(NodeEvaluator.java:657)
	at gyro.core.scope.NodeEvaluator.visitReference(NodeEvaluator.java:79)
	at gyro.lang.ast.value.ReferenceNode.accept(ReferenceNode.java:68)
	at gyro.lang.ast.NodeVisitor.visit(NodeVisitor.java:37)
	at gyro.core.directive.DirectiveProcessor.convert(DirectiveProcessor.java:109)
	at gyro.core.directive.DirectiveProcessor.getArgument(DirectiveProcessor.java:117)
	at gyro.core.PrintDirectiveProcessor.process(PrintDirectiveProcessor.java:32)
	at gyro.core.scope.NodeEvaluator.visitDirective(NodeEvaluator.java:344)
	at gyro.core.scope.NodeEvaluator.visitDirective(NodeEvaluator.java:79)
	at gyro.lang.ast.block.DirectiveNode.accept(DirectiveNode.java:85)
	at gyro.lang.ast.NodeVisitor.visit(NodeVisitor.java:37)
	at gyro.core.scope.NodeEvaluator.lambda$evaluateBody$30(NodeEvaluator.java:302)
	at gyro.core.scope.Defer.execute(Defer.java:50)
	at gyro.core.scope.NodeEvaluator.evaluateBody(NodeEvaluator.java:302)
	at gyro.core.scope.NodeEvaluator.visitFile(NodeEvaluator.java:385)
	at gyro.core.scope.NodeEvaluator.visitFile(NodeEvaluator.java:79)
	at gyro.lang.ast.block.FileNode.accept(FileNode.java:32)
	at gyro.lang.ast.NodeVisitor.visit(NodeVisitor.java:37)
	at gyro.core.scope.NodeEvaluator.lambda$evaluateBody$30(NodeEvaluator.java:302)
	at gyro.core.scope.Defer.execute(Defer.java:50)
	at gyro.core.scope.NodeEvaluator.evaluateBody(NodeEvaluator.java:302)
	at gyro.core.scope.NodeEvaluator.evaluate(NodeEvaluator.java:286)
	at gyro.core.scope.RootScope.evaluate(RootScope.java:313)
	at gyro.core.command.AbstractConfigCommand.doExecute(AbstractConfigCommand.java:132)
	at gyro.core.command.AbstractCommand.execute(AbstractCommand.java:71)
	at gyro.cli.Gyro.run(Gyro.java:192)
	at gyro.cli.Gyro.main(Gyro.java:76)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)


```

File: src/main/java/gyro/google/compute/ProjectMetadataItemResource.java (hunk 1 above)
Comment (alex-lambrides) [reply src/main/java/gyro/google/compute/ProjectMetadataItemResource.java]: The API actually handles the validation for us, and will throw an error like the following when the key doesn't match the regex: `Caused by: Error: Invalid value for field 'metadata.items[1].key': 'example-key**'. Must be a match of regex '[a-zA-Z0-9-_]{1,128}'`

File: src/main/java/gyro/google/compute/ProjectMetadataItemResource.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ *         value: "example-value"
+ *     end
+ */
+@Type("project-metadata-item")
+public class ProjectMetadataItemResource extends ComputeResource implements Copyable<Metadata.Items> {
+    private String key;
+    private String value;
+
+    /**
+     * The key of the metadata item. (Required)
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ProjectMetadataItemResource.java]: Even though we are not validating but we should mention the constrains for the `key` field

File: src/main/java/gyro/google/compute/ProjectMetadataItemResource.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+        Metadata metadata = getMetadata(client);
+        metadata.getItems().removeIf(r -> getKey().equals(r.getKey()));
+
+        setMetadata(client, metadata);
+    }
+
+    private Metadata getMetadata(Compute client) {
+        try {
+            Compute.Projects.Get projectRequest = client.projects().get(getProjectId());
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ProjectMetadataItemResource.java]: For consistency we should also catch the `GoogleJsonResponseException`

**Pull request merged**